src/types/pick.ts

```diff
@@ -0,0 +1,139 @@
+            if (missingKey) {
+                partialKey || details.push({ kind: 'missing property', property: key, type: innerType });
+                continue;
+            }
+
+            const innerResult = innerType.validate(input[key], options);
+            if (innerResult.ok) {
+                constructResult[key] = innerResult.value;
+            } else {
+                details.push(...prependPathToDetails(innerResult, key));
+            }
+        }
+        return this.createResult(input, options.mode === 'construct' ? constructResult : input, details);
+    }
+    override accept<R>(visitor: Visitor<R>): R {
+        return visitor.visitObjectLikeType(this);
+    }
+    /** {@inheritdoc BaseTypeImpl.maybeStringify} */
+    override maybeStringify(value: ResultType): string {
+        return interfaceStringify(this.propsArray, value as Record<string, unknown>);
+    }
+}
+define(PickType, 'basicType', 'object');
+
+export function pick<TypeImpl extends PickableImpl, Keys extends PickableKeys<TypeOf<TypeImpl>>>(
+    ...args: [name: string, baseType: TypeImpl, keys: OneOrMore<Keys>] | [baseType: TypeImpl, keys: OneOrMore<Keys>]
+): ObjectType<DistributedPick<TypeOf<TypeImpl>, Keys>> {
+    const [name, baseType, keys] = decodeOptionalName(args);
+    if (baseType instanceof UnionType) {
+        const unionType: UnionType<OneOrMore<BaseObjectLikeTypeImpl<TypeOf<TypeImpl>>>, TypeOf<TypeImpl>> = baseType;
+        if (!(unionType.basicType === 'object')) {
+            throw new Error(`Can only pick elements of unions with 'object' as basic type.`);
+        }
+        const narrowedKeys = narrowPickedKeys(keys, unionType.types);
+        if (!validPick(keys, narrowedKeys)) {
+            throw new Error('Selected keys describe impossible union variant.');
+        }
+
+        const innerTypes = unionType.types
+            .map((innerType, idx) => {
+                const specificKeys = narrowedKeys[idx] ?? [];
+                const keyCount = specificKeys.length;
+                // Only build the pick union elements if at least one of their keys has been picked.
+                // Also temporarily store the key count in a tuple to sort.
+                return [keyCount, keyCount ? new PickType(innerType, checkOneOrMore(specificKeys)) : undefined] as const;
+            })
+            // Sorting is necessary so the union type will internally match with a maximum subtype. This is so that types don't accidentally
+            // get narrowed too much during parsing and stringifying.
+            .sort(([aKeys, _1], [bKeys, _2]) => bKeys - aKeys) // Reverse sort: from most keys to least keys.
+            .flatMap(([_, innerType]) => innerType ?? []);
+
+        return union(
+            // Manually rename the type, because it's confusing if your union elements are suddenly in a different order.
+            name ?? `Pick<${baseType.name}, ${keys.map(key => `'${key}'`).join(' | ')}>`,
+            checkOneOrMore(innerTypes),
```

Comment:
Yeah so for example:
```ts
const A = object('A', { a: string });
const B = object('B', { a: string, b: boolean, d: undefined });
const C = object('C', { a: string, b: boolean, c: null  });
const OriginalUnion = union([A, B, C]);
// OriginalUnion.name === 'A | B | C'
cont PickedUnion = pick(OriginalUnion, ['a', 'b', 'd']);
// Because of the sorting, PickedUnion.name would be: 
// Pick<B, 'a' | 'b' | 'd'> | Pick<C, 'a' | 'b'> | Pick<A, 'a'>
// Not wrong per se, just a bit odd that the elements are suddenly in a different order

// With the current implementation, the name will look like this instead:
// Pick<A | B | C, 'a' | 'b' | 'd'>
```
Still a bit in doubt about this, as the first name does show the distributive property much clearer.